src/librustc/session/config.rs

```diff
@@ -933,7 +957,28 @@ pub fn build_session_options(matches: &getopts::Matches) -> Options {
         output_types.insert(OutputType::Exe, None);
     }
 
-    let cg = build_codegen_options(matches, color);
+    let mut cg = build_codegen_options(matches, color);
+
+    // Issue #30063: if user requests llvm-related output to one
+    // particular path, disable codegen-units.
+    if matches.opt_present("o") && cg.codegen_units != 1 {
```

Comment:
Okay; I was trying to minimize the impact of the change over the current state of affairs, but I'm willing to concede that the principle of least surprise implies we should adopt the tack you suggest.